### PR TITLE
Add centrifuge classification

### PR DESCRIPTION
**Pull request by sofstam**
This pull request includes addition of centrifuge module in `taxprofiler.nf`.

## PR checklist

- [x] This comment contains a description of changes (with reason).
- [ ] If you've fixed a bug or added code that should be tested, add tests!
    - [ ] If you've added a new tool - have you followed the pipeline conventions in the [contribution docs](https://github.com/nf-core/taxprofiler/tree/master/.github/CONTRIBUTING.md)
    - [x] If necessary, also make a PR on the nf-core/taxprofiler _branch_ on the [nf-core/test-datasets](https://github.com/nf-core/test-datasets) repository.
- [x] Make sure your code lints (`nf-core lint`).
- [x] Ensure the test suite passes (`nextflow run . -profile test,docker` --outdir <OUTDIR>`).
- [ ] Usage Documentation in `docs/usage.md` is updated.
- [ ] Output Documentation in `docs/output.md` is updated.
- [ ] `CHANGELOG.md` is updated.
- [ ] `README.md` is updated (including new tool citations and authors/contributors).

### CODE REVIEW

File: subworkflows/local/db_check.nf [hunk 1]
```diff
@@ -21,7 +21,7 @@ workflow DB_CHECK {
 
     ch_dbs_for_untar = parsed_samplesheet
         .branch {
-            untar: it[1].toString().endsWith(".tar.gz")
+            untar: it[1].toString().endsWith(".tar.gz") && it[0]['tool']!="centrifuge"
```
Comment (jfy133) [inline subworkflows/local/db_check.nf:24]:
```suggestion
            untar: it[1].toString().endsWith(".tar.gz") && it[0]['tool'] != "centrifuge"
```

File: nextflow.config [hunk 2]
```diff
@@ -65,6 +65,12 @@ params {
 
     // kraken2
     run_kraken2                = false
+
+    // centrifuge
+    run_centrifuge             = false
+    save_unaligned             = false
```
Comment (jfy133) [inline nextflow.config:71]:
```suggestion
    save_unaligned             = false
```

For all these parameters can you please append with `centrifuge_` so we know which tool it's relevant for.

**Review by jfy133 — COMMENTED** [review]
Looks really good @sofstam ! Only one change request re-paramaters and one formatting thing.

Before we merge, can you pull upstream from dev (to make sure you have the latest template).

Can you also run `nf-core schema build` to make sure the new parameters are in the schema. You don't have to fill in any of the information or help info, just make sure the parameter type is boolean. 

We will fill in all the documentation before the first release to make sure everything is consistent :+1:

**Review by jfy133 — APPROVED** [review]

**Comment by jfy133** [comment]
Just need the `nf-core schema build` update and we are good to go :)

File: modules/nf-core/modules/centrifuge/main.nf [hunk 3]
```diff
@@ -0,0 +1,63 @@
...
+    if (meta.single_end) {
+        unaligned = save_unaligned ? "--un-gz ${prefix}.unmapped.fastq.gz" : ''
+        aligned = save_aligned ? "--al-gz ${prefix}.mapped.fastq.gz" : ''
+    } else {
+        unaligned = save_unaligned ? "--un-conc-gz ${prefix}.unmapped.fastq.gz" : ''
+        aligned = save_aligned ? "--al-conc-gz ${prefix}.mapped.fastq.gz" : ''
+    }
+    def sam_output = sam_format ? "--out-fmt 'sam'" : ''
+    """
+    tar -xf $db
```
Comment (Midnighter) [inline modules/nf-core/modules/centrifuge/main.nf:45]: Extracting the database for every instance of this process seems sub-optimal. I suggest the database is extracted using the untar module and that this process is passed the extracted database directly.

**Review by Midnighter — COMMENTED** [review]
It seems to be a common pattern that these tools name their output `.txt` when in fact those are often tabular tab-separated values. I vote that in the pipeline in general we name something `.tsv` when it's in fact a TSV file.

**Comment by jfy133** [comment]
> It seems to be a common pattern that these tools name their output `.txt` when in fact those are often tabular tab-separated values. I vote that in the pipeline in general we name something `.tsv` when it's in fact a TSV file.

Agreed, this is also the purpose of the parsing step afterwords. Ideally we will publish those tables rather than the ones generated by the tools themselves

File: modules/nf-core/modules/centrifuge/main.nf (hunk 3 above)
Comment (jfy133) [reply modules/nf-core/modules/centrifuge/main.nf:45]: @sofstam is this now fixed by the tar repacking we talked about on slack? Will you update hte module?

**Comment by sofstam** [comment]
I updated the PR. I have a question regarding the following line in `nextflow.config`: 

   ```
 // centrifuge
    run_centrifuge             = false
    centrifuge_db_name         = 'minigut_cf'
    centrifuge_save_unaligned  = false
    centrifuge_save_aligned    = false
    centrifuge_sam_format      = false
```

The tests pass but I am wondering if I should pass a different value for `centrifuge_db_name`.

File: nextflow.config [hunk 4]
```diff
@@ -72,6 +72,12 @@ params {
     // kraken2
     run_kraken2                = false
 
+    // centrifuge
+    run_centrifuge             = false
+    centrifuge_db_name         = 'minigut_cf'
+    centrifuge_save_unaligned  = false
+    centrifuge_save_aligned    = false
+    centrifuge_sam_format      = false
     // metaphlan3
```
Comment (jfy133) [inline nextflow.config:81]:
```suggestion
    // metaphlan3
    
```

**Review by jfy133 — COMMENTED** [review]
Re you question:

We don't want to set this as a user-facing parameter at all for taxprofiler, because we store this information in the `database.csv` meta information.

You will need to use the `multiMap` option to pull this information out of `meta[]` somehow and insert it into the process input channel... I'm not sure how though unfortunately... might need to do some slightly hacky string manipulation to extract it from the db_meta... :\

**Comment by sofstam** [comment]
Thank you for your input. I will think of how this could be implemented tomorrow. Otherwise, maybe revise again the module.

File: conf/test.config [hunk 5]
```diff
@@ -25,9 +25,10 @@ params {
     input                         = 'https://raw.githubusercontent.com/nf-core/test-datasets/taxprofiler/samplesheet.csv'
     databases                     = 'https://raw.githubusercontent.com/nf-core/test-datasets/taxprofiler/database.csv'
     run_kraken2                   = true
-    run_malt                      = true
+    run_malt                      = false
```
Comment (jfy133) [inline conf/test.config:28]:
```suggestion
    run_malt                      = true
```

If you want to turn it off during testing, you can specify `-profile <container>,test --run_malt false` :)

File: conf/test.config [hunk 6]
```diff
@@ -25,9 +25,10 @@ params {
...
     run_metaphlan3                = true
+    run_centrifuge                = true
     shortread_clipmerge           = true
     longread_clip                 = false
-    shortread_complexityfilter    = true
+    shortread_complexityfilter    = false
```
Comment (jfy133) [inline conf/test.config:33]:
```suggestion
    shortread_complexityfilter    = true
```

File: modules/nf-core/modules/nf-core/modules/centrifuge/main.nf [hunk 7]
```diff
@@ -0,0 +1,63 @@
+process CENTRIFUGE {
```
Comment (jfy133) [inline modules/nf-core/modules/nf-core/modules/centrifuge/main.nf:1]: This file should be removed now? `nf-core modules remove centrifuge` I think...

File: modules/nf-core/modules/nf-core/modules/centrifuge/meta.yml [hunk 8]
```diff
@@ -0,0 +1,73 @@
+name: centrifuge
```
Comment (jfy133) [inline modules/nf-core/modules/nf-core/modules/centrifuge/meta.yml:1]: Same here, but should be removed with the previous command

File: subworkflows/local/input_check.nf [hunk 9]
```diff
@@ -63,13 +63,12 @@ def create_fastq_channel(LinkedHashMap row) {
...
-            fastq_meta = [ meta, [ file(row.fastq_1), file(row.fastq_2) ] ]
+         fastq_meta = [ meta, [ file(row.fastq_1), file(row.fastq_2) ] ]
         }
+
     }
     return fastq_meta
-}
-
-// Function to get list of [ meta, fasta ]
+}// Function to get list of [ meta, fasta ]
```
Comment (jfy133) [inline subworkflows/local/input_check.nf:71]: this looks a bit funky? Is the change corret?

**Review by jfy133 — CHANGES_REQUESTED** [review]
A bit of clean up is required but overall looking pretty good! I think you're almost freeee from centrifuge!

File: conf/test.config (hunk 5 above)
Comment (sofstam) [reply conf/test.config:28]: Sorry, my bad :)

File: subworkflows/local/input_check.nf (hunk 9 above)
Comment (sofstam) [reply subworkflows/local/input_check.nf:71]: I had no intention of changing it, I think this is a typo.

**Comment by sofstam** [comment]
I was thinking of working with the second subworkflow for centrifuge today and change this PR accordingly.

**Comment by jfy133** [comment]
> I was thinking of working with the second subworkflow for centrifuge today and change this PR accordingly.

I would suggest we get this in already, as the second step is just for generating a differently formattged profile I believe?

So I once you remove the wierd module install (see screesnhot), and tests pass, I will move already and you can make a second PR for when the kreport module is ready :+1:  

image

**Comment by sofstam** [comment]
Sounds good to me, would be great to see centrifuge being merged in dev 🥳

**Review by jfy133 — APPROVED** [review]

**Pull request merged**